scripts/chat.py

```diff
@@ -0,0 +1,276 @@
+    group=dist.group.WORLD,
+)
+tokenizer = tokenizers.get_tokenizer(args.tokenizer)
+model.eval()
+torch.set_grad_enabled(False)
+print("loading complete on rank", local_rank)
+
+if args.compile:
+    print("compiling model")
+    # Bug with kv-cache in PT2.1
+    torch._inductor.config.joint_graph_constant_folding = False
+    # compiling can make first inference pass slow
+    model = torch.compile(model, mode=args.compile_mode)
+
+
+"""
+llama template for chat looks like:
+
+<bos>[INST]<<SYS>>system message<</SYS>>Instruction prompt[/INST]Answer<eos><bos>[INST] Prompt [/INST] Answer <eos><bos>[INST] Prompt [/INST] Answer <eos>
+
+specified here:
+https://github.com/huggingface/transformers/blob/v4.40.2/src/transformers/models/llama/tokenization_llama.py#L438
+
+and here:
+https://github.com/meta-llama/llama/blob/556949fdfb72da27c2f4a40b7f0e4cf0b8153a28/llama/generation.py#L320-L362
+
+TODO: add some way to specify alternate templates
+"""
+
+
+class ChatREPL(cmd.Cmd):
+    def __init__(self, max_new_tokens, system_message=None):
+        super().__init__()
+        self.intro = 'Enter your prompts/questions or "/help" for usage instructions.\n'
+        self.prompt = "$ "
+        self.B_SYS = "<<SYS>>\n"
+        self.E_SYS = "\n<<SYS>>\n\n"
+        self.B_INST = "[INST]"
+        self.E_INST = "[/INST]"
+        self.max_new_tokens = max_new_tokens
+
+        self.buffer = []
+        self.buffer.append(tokenizer.bos_token_id)
+        if system_message:
+            msg = self.B_SYS + system_message + self.E_SYS
+            self.buffer.extend(tokenizer.convert_tokens_to_ids(msg))
+
+    def onecmd(self, line):
+        # Check if the line starts with a special command prefix
+        if line.startswith("/"):
+            command = line[1:].strip().split(" ")[0]
+            args = line[len(command) + 2 :].strip()
+            try:
+                func = getattr(self, "do_" + command)
+                return func(args)
+            except AttributeError:
+                print(f"Unknown command: {line}")
+                return
+        # handles ctrl-D
+        elif line == "EOF":
+            self.do_EOF()
+        else:
+            return self.default(line)
+
+    def default(self, line):
+        if not len(line.strip()):
+            return
+
+        self.buffer.extend(
+            tokenizer.convert_tokens_to_ids(
+                tokenizer.tokenize(self.B_INST + line + self.E_INST)
+            )
+        )
+        ids = torch.tensor(self.buffer, dtype=torch.long, device=device)
+
+        result = generate(
+            model,
+            ids,
+            max_new_tokens=self.max_new_tokens,
+            use_cache=args.no_use_cache,
+            do_sample=not args.deterministic,
+            eos_token_id=tokenizer.eos_token_id,
+        )
+        if local_rank != 0:
+            return
+
+        result = result[len(self.buffer) :].tolist()
+        self.buffer = self.buffer + result + [tokenizer.eos_token_id]
```

Comment:
how does this interact with the changes in generation about finishing on seeing EOS tokens?